**配置环境**：
```sh
sudo apt-get install axel
axel -n 16 -c https://mirrors.tuna.tsinghua.edu.cn/anaconda/miniconda/Miniconda3-latest-Linux-x86_64.sh
sh Miniconda3-latest-Linux-x86_64.sh
source ~/.bashrc
conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/free/
conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge/
conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/bioconda/
conda config --set show_channel_urls yes
conda config --add channels https://mirrors.ustc.edu.cn/anaconda/pkgs/free/
conda config --add channels https://mirrors.ustc.edu.cn/anaconda/cloud/conda-forge/
conda config --add channels https://mirrors.ustc.edu.cn/anaconda/cloud/bioconda/
conda config --set show_channel_urls yes
conda create -n download -y \
    sra-tools parallel-fastq-dump
conda create -n STAR1 -y \
    star=2.7.0 cutadapt=1.9 parallel
conda create -n STAR2 -y \
    star=2.7.1 samtools=1.9 deeptools=2.2 stringtie=2.1.4 taco=0.7.3 parallel
conda create -n R -y \
    r-base=4.4 r-data.table
# 验证R环境安装
conda activate R
R  # 进入R环境
install.packages("optparse")
install.packages("data.table")
install.packages("magrittr")
install.packages("reshape")
install.packages("BiocManager")
library("BiocManager")
options(BioC_mirror = "https://mirrors.westlake.edu.cn/bioconductor")
BiocManager::install("XVector")
BiocManager::install("GenomicRanges")
library(optparse)
library(parallel)
library(methods)
library(data.table)
library(magrittr)
library(GenomicRanges)
library(reshape)
conda deactivate
```

安装XVector/GenomicRanges包时报错`io_utils.c:16:10: fatal error: zlib.h: No such file or directory`
```sh
cd /home/lwl
mkdir zlib && cd zlib
wget https://zlib.net/zlib-1.3.1.tar.gz
tar -zxvf zlib-1.3.1.tar.gz 
cd zlib-1.3.1
./configure --prefix=$HOME/zlib-1.3.1
make
make install  # 应该会在$HOME下生成一个zlib-1.3.1文件夹
sudo nano ~/.bashrc
# 添加以下内容后
source ~/.bashrc
```

```
export PATH=$PATH:$HOME/zlib-1.3.1/ 
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$HOME/zlib-1.3.1/lib/
export LIBRARY_PATH=$LIBRARY_PATH:$HOME/zlib-1.3.1/lib/
export C_INCLUDE_PATH=$HOME/zlib-1.3.1/include/
export CPLUS_INCLUDE_PATH=$HOME/zlib-1.3.1/include/
export PKG_CONFIG_PATH=$HOME//zlib-1.3.1/lib/pkgconfig
```

**样本下载**：以SRR521514为例
```sh
threads=16
baseDir=/home/user
SRR=SRR521514
conda activate download
mkdir fastq && cd fastq
parallel-fastq-dump --sra-id ${SRR} --threads ${threads} --split-files
gzip ${SRR}_1.fastq && mv ${SRR}_1.fastq.gz ${SRR}_R1.fastq.gz
gzip ${SRR}_2.fastq && mv ${SRR}_2.fastq.gz ${SRR}_R2.fastq.gz
conda deactivate
```

**STAR建索引**：
```sh
conda activate STAR1
mkdir refs && cd refs
mkdir STAR_refs
axel -n ${threads} -c https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_49/GRCh38.p14.genome.fa.gz
axel -n ${threads} -c https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_49/gencode.v49.annotation.gtf.gz
gunzip GRCh38.p14.genome.fa.gz
gunzip gencode.v49.annotation.gtf.gz
STAR --runMode genomeGenerate \
     --runThreadN ${threads} \
     --genomeDir STAR_refs \
     --genomeFastaFiles GRCh38.p14.genome.fa \
     --sjdbGTFfile gencode.v49.annotation.gtf
genomeDir=${baseDir}/refs/STAR_refs
STAR --genomeDir ${genomeDir} --genomeLoad LoadAndExit 
# rmsk注释
mkdir rmsk_refs && cd rmsk_refs
axel -n ${threads} -c https://hgdownload.soe.ucsc.edu/goldenPath/hg38/database/rmsk.txt.gz
gunzip rmsk.txt.gz
echo -e "chr\tstart\tend\tstrand\trepName\trepClass\trepFamily" > hg38.rmsk.cleaned.txt
cut -f6-8,10,11,12,13 rmsk.txt > hg38.rmsk.cleaned.txt
```
**Step 1: Adapter Trimming and First Pass Alignment**：
STAR Version: STAR-2.7.0c
Cutadapt Version: 1.9
需要先去`first_alignment_PE_TruSeq.sh`修改软件存放路径
```sh
mkdir TE_Transcript_Assembly && cd TE_Transcript_Assembly
mkdir scripts  # 脚本文件放这
parallel -uj 10 ./scripts/first_alignment_PE_TruSeq.sh ${SRR} ${genomeDir} 5
conda deactivate
```
