```sh
mkdir ERVmap
cd ERVmap
mkdir input
parallel-fastq-dump --sra-id SRR521514 --threads 8 --split-files --outdir input
parallel-fastq-dump --sra-id SRR521515 --threads 8 --split-files --outdir input
cat SRR521514_1.fastq SRR521515_1.fastq > H1_R1.fastq
cat SRR521514_2.fastq SRR521515_2.fastq > H1_R2.fastq
gzip H1_R1.fastq
gzip H1_R2.fastq
```

```sh
sudo apt-get update && sudo apt-get install -y docker.io
sudo usermod -aG docker $USER  # 重启终端
cd /etc/docker
sudo nano daemon.json
```
```
{ "registry-mirrors": ["https://docker.registry.cyou", 
"https://docker-cf.registry.cyou", "https://dockercf.jsdelivr.fyi", 
"https://docker.jsdelivr.fyi", "https://dockertest.jsdelivr.fyi", 
"https://mirror.aliyuncs.com", "https://dockerproxy.com", 
"https://mirror.baidubce.com", "https://docker.m.daocloud.io", 
"https://docker.nju.edu.cn", 
"https://docker.mirrors.sjtug.sjtu.edu.cn", 
"https://docker.mirrors.ustc.edu.cn", "https://mirror.iscas.ac.cn", 
"https://docker.rainbond.cc"]
}
```

```sh
docker pull eipm/ervmap
```

https://hub.docker.com/r/eipm/ervmap

```sh
wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_49/GRCh38.p14.genome.fa.gz
wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_49/gencode.v49.annotation.gtf.gz
for gz in *.gz; do gunzip $gz; done
STAR --runMode genomeGenerate \
     --runThreadN 8 \
     --genomeDir  ~/ervmap/ref/STAR_hg38 \
     --genomeFastaFiles GRCh38.p14.genome.fa \
     --sjdbGTFfile      gencode.v49.annotation.gtf \
     --sjdbOverhang 99 \
     --limitGenomeGenerateRAM 125000000000
```

```sh
docker run --rm -it \
  -v ~/ervmap/input:/work/input \
  -v ~/ervmap/ref:/work/ref \
  -v ~/ervmap/out:/work/output \
  eipm/ervmap \
  bash -lc "cd /opt/ERVmap && bash ERVmap_auto.sh /work/input"
```
